sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/PerSubscriptionPartitionSdf.java

```diff
@@ -81,15 +84,22 @@ public ProcessContinuation processElement(
           .lastClaimed()
           .ifPresent(
               lastClaimedOffset ->
-                  finalizer.afterBundleCommit(
-                      Instant.ofEpochMilli(Long.MAX_VALUE),
-                      () -> {
-                        Committer committer = committerFactory.apply(subscriptionPartition);
-                        committer.startAsync().awaitRunning();
-                        // Commit the next-to-deliver offset.
-                        committer.commitOffset(Offset.of(lastClaimedOffset.value() + 1)).get();
-                        committer.stopAsync().awaitTerminated();
-                      }));
+              /* TODO(boyuanzz): When default dataflow can use finalizers, undo this.
+              finalizer.afterBundleCommit(
+                  Instant.ofEpochMilli(Long.MAX_VALUE),
+                  () -> */ {
+                Committer committer = committerFactory.apply(subscriptionPartition);
+                committer.startAsync().awaitRunning();
+                // Commit the next-to-deliver offset.
+                try {
+                  committer.commitOffset(Offset.of(lastClaimedOffset.value() + 1)).get();
```

Comment:
There are several concerns around committing offset in the `DoFn`:

- It's possible that Dataflow fails to process one element and re-process it again. In this case, one offset might be committed for more than once.
- It's also possible that one offset is committed successfully but the pipeline crashes before downstream actually  consumes the output. That might result in data loss.

If these concerns are not true for your source, then that's ok to do so.

Reply:
1) I don't know that there's a way to prevent reprocessing without finalizer. Even the kafka method (commits in a PTransform afterwards) can do this, as PCollections are unordered so you can potentially (even if unlikely) commit in opposite orders, leading to backwards commits. This, if I understand the execution model correctly, can lead to duplicate commits only in the case of an internal dataflow failure, in which case there is spurious progress.

2) For "Pipeline crashes", do you mean the entire dataflow job can fail or the downstream processing just does `throw new RuntimeException()`? In the second, doesn't BundleFinalizer also have this issue? If the first... I'm not sure its worth protecting something in the runner against runner incorrectness.

Reply:
For both 1&2, as long as the committing offset doesn't means we are no longer able to read from that offset, I think that would be fine.